modules/core/src/mathfuncs.cpp

```diff
@@ -2511,12 +2509,46 @@ double cv::solvePoly( InputArray _coeffs0, OutputArray _roots0, int maxIters )
         {
             p = roots[i];
             C num = coeffs[n], denom = coeffs[n];
+            int num_same_root = 1;
             for( j = 0; j < n; j++ )
             {
                 num = num*p + coeffs[n-j-1];
-                if( j != i ) denom = denom * (p - roots[j]);
+                if( j != i )
+                {
+                    if ( (p - roots[j]).re != 0 || (p - roots[j]).im != 0 )
+                        denom = denom * (p - roots[j]);
+                    else
+                        num_same_root++;
+                }
             }
             num /= denom;
+            if( num_same_root > 1)
+            {
+                for( j = 0; j < num_same_root / 2; j++)
+                {
+                    double old_num_re = num.re;
+                    double old_num_im = num.im;
+
+                    num.re = old_num_re*old_num_re + old_num_im*old_num_im;
+                    Mat num_re_arr = (Mat_<double>(1,1) << num.re);
+                    cv::sqrt(num_re_arr, num_re_arr);
```

Comment:
this is terribly inefficient - allocate 1x1 matrix and call cv::sqrt(). just use std::sqrt() instead